File: packages/ai_config/discovery/discovery_engine.py
from typing import Dict, List


def run_provider_discovery(plugins: List, context: Dict | None = None) -> Dict:
    context = context or {}
    ordered = sorted(plugins, key=lambda plugin: (plugin.order, plugin.id))

    providers = []
    models = []

    for plugin in ordered:
        result = plugin.discover(context)
        providers.extend(result.get("providers", []))
        models.extend(result.get("models", []))

    return {"providers": providers, "models": models}

File: packages/ai-config/src/discovery/discoveryEngine.ts
import type { Catalog } from '../domain/catalog.js';
import type { ProviderDiscoveryContext, ProviderDiscoveryPlugin } from './pluginInterface.js';

export async function runProviderDiscovery(
  plugins: ProviderDiscoveryPlugin[],
  context: ProviderDiscoveryContext = {},
): Promise<Partial<Catalog>> {
  const ordered = [...plugins].sort((a, b) => a.order - b.order || a.id.localeCompare(b.id));
  const aggregate: Catalog = { providers: [], models: [] };

  for (const plugin of ordered) {
    const result = await plugin.discover(context);
    if (result.providers) {
      aggregate.providers.push(...result.providers);
    }
    if (result.models) {
      aggregate.models.push(...result.models);
    }
  }

  return aggregate;
}
